ssl_connect.c

```diff
@@ -0,0 +1,180 @@
+        {
+            sprintf(ssl_saved_error, "Failed to load root certificates\n");
+            ssl_error_type = 1;
+            err = true;
+            break;
+        }
+
+        if (SSL_CTX_use_certificate_file(ctx, client_certificate, SSL_FILETYPE_PEM) != 1)
+        {
+            sprintf(ssl_saved_error, "Failed to Load the client certificate\n");
+            ssl_error_type = 2;
+            err = true;
+            break;
+        }
+
+        if (SSL_CTX_use_PrivateKey_file(ctx, client_key, SSL_FILETYPE_PEM) != 1)
+        {
+            sprintf(ssl_saved_error, "Failed to Load the client key\n");
+            ssl_error_type = 2;
+            err = true;
+            break;
+        }
+    } while (0);
+
+    if (err)
+    {
+        ssl_error = ERR_get_error();
+        SSL_CTX_free(ctx);
+        ctx = 0;
+    }
+
+    return ctx;
+}
+
+const char* openssl_strerror()
+{
+    static int error_strings_loaded = 0;
+
+    if (!error_strings_loaded)
+    {
+        SSL_load_error_strings();
+        error_strings_loaded = 1;
+    }
+	return ERR_error_string(ERR_get_error(), NULL);
+}
+
+//------------------------------------------------------------------------------
+BIO* ssl_connect(SSL_CTX *ctx, const char* server_address, const char* server_port, SSL_SESSION** session)
+{
+    /* Set up the TLS connection to the KMIP server. */
+    SSL* ssl = 0;
+    BIO* bio = 0;
+    ssl_error = 0;
+
+    bio = BIO_new_ssl_connect(ctx);
+    if(bio == NULL)
+    {
+        sprintf(ssl_saved_error, "BIO_new_ssl_connect failed: %s\n", openssl_strerror());
+        ssl_error = ERR_get_error();
+        return (0);
+    }
+
+    BIO_set_conn_hostname(bio, server_address);
+    BIO_set_conn_port(bio, server_port);
+    BIO_get_ssl(bio, &ssl);
+    SSL_set_mode(ssl, SSL_MODE_AUTO_RETRY);
+    if (*session)
+        SSL_set_session(ssl, *session);
+
+    if (BIO_do_connect(bio) != 1)
+    {
+        const char* certerr = NULL;
+        int rc = ERR_peek_error();
+        if ( rc )
+        {
+            if ( ( ERR_GET_LIB(rc) == ERR_LIB_SSL ) &&
+                 ( ERR_GET_REASON(rc) == SSL_R_CERTIFICATE_VERIFY_FAILED ) )
+            {
+                int certrc = SSL_get_verify_result(ssl);
+                certerr = (char *)X509_verify_cert_error_string(certrc);
+            }
+        }
+
+        ssl_error = rc;
+        sprintf(ssl_saved_error,"BIO_do_connect failed: %s\n", openssl_strerror());
+        if (certerr)
+            sprintf(ssl_saved_error, "SSL cert error: %s\n", certerr );
+
+        BIO_free_all(bio);
+        return(0);
+    }
+
+
+    if (SSL_session_reused(ssl)) {
+        ssl_reused = 1;
+        // printf("REUSED SESSION\n");
+    } else {
+        ssl_reused = 0;
+        // printf("NEW SESSION\n");
+    }
+
+    if (*session)
+        SSL_SESSION_free(*session);
+
+    *session = SSL_get1_session(ssl);
+
+    return bio;
+}
+
+void ssl_disconnect(BIO* bio)
+{
+    BIO_ssl_shutdown(bio);
+    BIO_free_all(bio);
+}
+
+
```

Comment:
For now separate out the Query changes from the SSL modifications. They should be evaluated separately.

Reply:
removed ssl wrapper